sidedish/src/component/Carousel/Carousel.jsx

```diff
@@ -1,61 +1,109 @@
-import React, { useRef, useState } from 'react';
+import React, { forwardRef, useEffect, useImperativeHandle, useRef, useState } from 'react';
 import styled from 'styled-components';
 import { IoChevronBackSharp, IoChevronForwardSharp } from 'react-icons/io5';
+import CarouselItem from './CarouselItem';
+import CarouselPage from './CarouselPage';
 
-const Carousel = ({ children, itemWidth, maxItem, skipItem, animationTime }) => {
-  const [locationX, setLocationX] = useState(0);
-  const [currIdx, setCurrIdx] = useState(0);
-  const [leftItem, setLeftItem] = useState();
+const Carousel = forwardRef(
+  (
+    { children, slideToScroll, speed, defaultArrow, defaultPaging, PagingComp, setCurrentIndex },
+    ref
+  ) => {
+    const [locationX, setLocationX] = useState(0);
+    const [currIdx, setCurrIdx] = useState(0);
+    const [leftItem, setLeftItem] = useState();
+    const [itemWidth, setItemWidth] = useState(); //308
+    const carouselContainerRef = useRef();
+    const isCarouselWidth = () => carouselContainerRef.current && itemWidth;
 
-  const handleClickPrev = () => {
-    const possibleMove = currIdx >= skipItem ? skipItem : currIdx;
-    setLocationX(locationX + itemWidth * possibleMove);
-    setCurrIdx(currIdx - possibleMove);
-    setLeftItem(leftItem + possibleMove);
-  };
+    const containerWidth = isCarouselWidth() && carouselContainerRef.current.offsetWidth; //1280
+    const slideToShow = isCarouselWidth() && Math.floor(containerWidth / itemWidth); //4
+    const marginRigthForItem =
+      isCarouselWidth() && (containerWidth - slideToShow * itemWidth) / (slideToShow - 1);
 
-  const handleClickNext = () => {
-    const totalItemCount = children.length;
-    const newLeftItem = totalItemCount - (currIdx + maxItem);
-    const possibleMove = newLeftItem >= skipItem ? skipItem : newLeftItem;
-    setLocationX(locationX - itemWidth * possibleMove);
-    setCurrIdx(currIdx + possibleMove);
-    setLeftItem(newLeftItem - possibleMove);
-  };
+    useEffect(() => {
+      setCurrentIndex && setCurrentIndex(slideToShow);
```

Comment:
```suggestion
      setCurrentIndex?.(slideToShow);
```

함수 호출도 optional chain으로 처리할 수 있답니다
https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Operators/Optional_chaining#%EB%AC%B8%EB%B2%95

Reply:
저희가 썼던 방법도 엄청 코드가 가독성이 좋다고 느꼈는데,
더 보기 좋은 방법이 있는 지는 몰랐네요. ㅎㅎㅎ 
앞으로 자주 써먹을 것 같습니다. 
감사합니다!!